MUTATION-SCOPE.md

```diff
@@ -0,0 +1,439 @@
+    // do some work that 'child' tracks
+    console.log("doing some nested work...")
+    // the nested span is closed when it's out of scope
+  }
+  await anotherWorkPromise;
+  // This parent span is also closed when it goes out of scope
+}
+```
+
+> This example is adapted from the OpenTelemetry Python example.
+> https://opentelemetry.io/docs/languages/python/instrumentation/#creating-spans
+
+Each `tracer.startAsCurrentSpan` invocation retrieves the parent span from its
+own `AsyncContext.Variable` instance and create span as a child, and set the
+child span as the current value of the `AsyncContext.Variable` instance:
+
+```js
+class Tracer {
+  #var = new AsyncContext.Variable();
+
+  startAsCurrentSpan(name) {
+    let scope;
+    const span = {
+      name,
+      parent: this.#var.get(),
+      [Symbol.enter]: () => {
+        scope = this.#var.withValue(span)[Symbol.enter]();
+        return span;
+      },
+      [Symbol.dispose]: () => {
+        scope[Symbol.dispose]();
+      },
+    };
+    return span;
+  }
+}
+```
+
+The set semantic that doesn't mutate existing snapshots is crucial to the
+`startAsCurrentSpan` example here, as it allows deferred span created in
+`doAnotherWork` to be a child span of the `"parent"` instead of `"child"`,
+shown as graph below:
+
+```
+⌌----------⌍
+| 'parent' |
+⌎----------⌏
+  |   ⌌---------⌍
+  |---| 'child' |
+  |   ⌎---------⌏
+  |   ⌌-----------------⌍
+  |---| 'doAnotherWork' |
+  |   ⌎-----------------⌏
+```
+
+### Alternative: Decouple mutation with scopes
+
+To preserve the strong scope guarantees provided by `run`, an additional
+constraint can also be put to `set` to declare explicit scopes of mutation.
+
+A dedicated `AsyncContext.contextScope` can be decoupled with `run` to open a
+mutable scope with a series of `set` operations.
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
+
+// Executes the `main` function in a new mutable context scope.
+AsyncContext.contextScope(() => {
+  asyncVar.set("main");
+
+  console.log(asyncVar.get()); // => "main"
+});
+// Goes out of scope and all variables are restored in the current context.
+
+console.log(asyncVar.get()); // => "default"
+```
+
+`AsyncContext.contextScope` is basically a shortcut of
+`AsyncContext.Snapshot.run`:
+
+```js
+const asyncVar = new AsyncContext.Variable({ defaultValue: "default" });
+
+asyncVar.set("A"); // Throws ReferenceError: Not in a mutable context scope.
+
+// Executes the `main` function in a new mutable context scope.
+AsyncContext.Snapshot.wrap(() => {
+  asyncVar.set("main");
+
+  console.log(asyncVar.get()); // => "main"
+})();
+// Goes out of scope and all variables are restored in the current context.
+
+console.log(asyncVar.get()); // => "default"
+```
+
+#### Use cases
+
+One use case of `set` is that it allows more intuitive test framework
```

Comment:
FWIW, I have a proof-of-concept for implementing `aroundEach` in Jasmine in userland entirely in terms of `beforeEach` and `afterEach`, though it relies on the legacy `done` style of async tests, so it wouldn't work in a more modern test framework that doesn't expose that.